pytorch_lightning/utilities/apply_func.py

```diff
@@ -129,8 +129,9 @@ def apply_to_collection(
                 **kwargs,
             )
             if include_none or v is not None:
-                out_dict[field] = v
-        return elem_type(**out_dict)
+                setattr(result, field.name, v)
+            # else retain old field value
```

Comment:
What should the expected behavior be here?

In other branches of `apply_to_collection`, `include_none=False` leads to the `None`-mapped item being dropped from the collection. I think it would be bad to drop a field from a dataclass, but setting it to `None` might be acceptable (letting it be user error if this violates some invariant for their dataclass).

Alternatively, the original value is retained as it is currently (as though the function was not applied to it).
At the moment, the function will retain the old field value such that `id(getattr(data, field.name)) == id(getattr(result, field.name))`, which seems problematic to me and could lead to unexpected mutability.
I think the proper fix would be to `deepcopy` the field value such that `id(getattr(data, field.name)) != id(getattr(result, field.name))` but `getattr(data, field.name) == getattr(result, field.name)`.

Let me know your preference.